File: 121.买卖股票的最佳时机.java
/*
 * @lc app=leetcode.cn id=121 lang=java
 *
 * [121] 买卖股票的最佳时机
 */

// @lc code=start
class Solution {
    // public int maxProfit(int[] prices) {
    //     // 前 i 天的最大收益 = max(前 i-1 天的最大收益，第 i 天的价格 - 前 i-1 天中的最小价格)
    //     int ans = 0, min = prices[0];
    //     for (int i = 1; i < prices.length; ++i) {
    //         ans = Math.max(ans, prices[i] - min);
    //         min = Math.min(min, prices[i]);
    //     }
    //     return ans;
    // }

    // Made similar to Q53: do maxSubArray() on the array of interday price differences
    public int maxProfit(int[] prices) {
        int ans = 0, dp = 0;
        for (int i = 1; i < prices.length; ++i) {
            int diff = prices[i] - prices[i - 1];
            ans = Math.max(ans, dp = Math.max(dp + diff, diff));
        }
        return ans;
    }
}
// @lc code=end

File: 121.买卖股票的最佳时机.py
#
# @lc app=leetcode.cn id=121 lang=python3
#
# [121] 买卖股票的最佳时机
#

# @lc code=start
class Solution:
    '''
    def maxProfit(self, prices: List[int]) -> int:
        # A more general solution: mono-stack,
        # which can be used in any problem of
        # the following form:
        # 用O(n)的时间得知所有位置两边第一个比他大(或小)的数的位置
        ans, stack = 0, []
        prices = prices + [-1]  # 加"哨兵"，确保单调栈中的每个元素都被进行判定
        for price in prices:
            while stack and stack[-1] > price:
                ans = max(ans, stack[-1] - stack[0])
                stack.pop()
            stack.append(price)
        return ans
    '''
    def maxProfit(self, prices: List[int]) -> int:
        min_price = float('inf')
        ans = 0
        for price in prices:
            ans = max(ans, price-min_price)
            min_price = min(min_price, price)
        return ans
    

# @lc code=end

